### PR TITLE
async: Introduce async indexing

### PR DESCRIPTION
**Pull request by asdine**
### What's being changed:

This PR adds support for asynchronous / batch indexing.
The idea is to allow clients to insert data as fast as possible without making them block until the data is indexed. They can then query their data before it's fully indexed, at the cost of query speed and accuracy.
We introduced an index queue that receives batches from the client, keeps them in-memory and distributes them to background indexing workers.
We also introduced a change in the HNSW index, the `AddBatch` method, which groups several inserts into one single write in the commit log.

This feature is behind a feature flag (`WEAVIATE_ASYNC_INDEXING=true`) and will mostly be tested internally for a while until we are confident it can replace the current behavior.
There are of course some corner cases that will need to be fine tuned and we are not expecting this first iteration to be the final solution.

### Sift

| Method                  | Import Time     | Indexing Time  |
|-------------------------|-----------------|----------------|
| Sync                    | 6m02s           |                |
| Async  | 1m39s  | 5m36s  |

### Sphere 1M

| Method                  | Import Time       | Indexing Time     |
|-------------------------|-------------------|-------------------|
| Sync                    | 33m14s            | -                 |
| Async | 9m35s | 26m32s |

### Before:
Currently on master, we have N main workers that are used by all the local shards to index vectors individually.
- the client sends an insert request (batch or not)
- we store the vectors in the LSM store
- each shard distributes each vector across the available workers for indexing
- we wait for all the vectors to be indexed + geoindexed (if enabled)
- we send a response to the client, with an error message for each vector, if any

### After:
Each shard has its own index queue. Each queue has its own set of dedicated workers.
- the client sends an insert request (batch or not)
- we store the vectors in the LSM store + (geoindex if enabled)
- the index queue stores the vectors in memory
- we send a response to the client
- the queue distributes chunks of the in-memory queue to its background workers (`BATCH_SIZE` parameter, default: 1000)
- each worker calls `AddBatch`

## 🔍 Search

For search queries, we merge the results of HNSW with the results of Brute force on the pending non-indexed vectors.
We limit the number of vectors passed to the Brute force algorithm (`BruteForceSearchLimit` parameter, defaul: 100k), to allow keeping a large number of vectors in memory without sacrificing too much query speed.
Once all vectors are indexed, we are expecting similar performance as before.

## ❌ Error handling

During an insert request, the client is blocking at least until the validation and persistence is done.
However, since indexing is happening in the background, we can't notify the client if an error happens.
In this PR, if an indexing error happens we simply retry indefinitely, until the indexing succeeds.
This will make the queue resilient to transient errors, but might block a worker indefinitely if the error is unrecoverable.

## 🚑  Crash recovery / startup

We keep track of indexed vectors using an optimistic approach and store that information in a BoltDB fille `index.db`.
In case of a restart or a crash, we simply iterate over objects in the LSM store, starting from that checkpoint, and enqueue vectors that are not found in the index.

##  🔥  Impact on memory / CPU

We should of course expect an increase in memory usage during imports since the queue is in-memory. There is currently no max queue size mechanism or back-pressure strategy.
With this design, it would be the user's responsibility to ensure they have enough ram to host their dataset.

## ⚙️  No configuration yet

At this stage, we don't expose any configuration to the user.
We will do so once the design is proven to be solid and we determine the right API.

## 🔬 PQ

Tested with PQ via chaos engineering pipelines

## TODO

- [x] Base implementation
- [x] Load previous vectors at startup
- [x] Search both unindexed/pending and indexed vectors
- [x] Handle deletion
- [x] Max number of vectors for brute force search 
- [x] Fix integration tests

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [x] All new code is covered by tests where it is reasonable.
- [x] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/vector/hnsw/index.go [hunk 1]
```diff
@@ -24,15 +24,19 @@ import (
...
 	"github.com/weaviate/weaviate/adapters/repos/db/vector/ssdhelpers"
 	"github.com/weaviate/weaviate/entities/cyclemanager"
 	"github.com/weaviate/weaviate/entities/storobj"
 	ent "github.com/weaviate/weaviate/entities/vectorindex/hnsw"
 )
 
-const NodeLockStripe = uint64(512)
+const (
+	NodeLockStripe = uint64(512)
+	TempSize       = 1000
```
Comment (trengrj) [inline adapters/repos/db/vector/hnsw/index.go:37]: Are `TempSize` and `TempWorkers` used?

Comment (asdine) [reply adapters/repos/db/vector/hnsw/index.go:37]: No they are not, I'm removing them

File: adapters/repos/db/index_queue_test.go [hunk 2]
```diff
@@ -0,0 +1,723 @@
...
+func newCheckpointManagerWithDir(t testing.TB, dir string) *indexcheckpoint.Checkpoints {
+	t.Helper()
+
+	c, err := indexcheckpoint.New(dir, logrus.New())
+	require.NoError(t, err)
+
+	return c
+}
+
+func pushVector(t testing.TB, ctx context.Context, q *IndexQueue, id uint64, vector ...float32) {
```
Comment (aliszka) [inline adapters/repos/db/index_queue_test.go:69]: could you change  type of `vector` to `[]float32`? now it kinda blends with `id` when used (`pushVector(t, ctx, q, 1, 1, 2, 3)`)

to have `pushVector(t, ctx, q, 1, []float32{1, 2, 3})`,
or `pushVector(t, ctx, q, 1, vec(1, 2, 3))` (`vec` function needs to be created in that case)

File: adapters/repos/db/vector/hnsw/deserializer.go [hunk 3]
```diff
@@ -393,17 +401,6 @@ func (d *Deserializer) ReadClearLinksAtLevel(r io.Reader, res *DeserializationRe
 		res.Nodes = newNodes
 	}
 
-	if keepReplaceInfo {
```
[the remark this answers is not in the document]
Comment (asdine) [reply adapters/repos/db/vector/hnsw/deserializer.go:396]: I'll cc @abdelr on this one

File: adapters/repos/db/index_queue.go [hunk 4]
```diff
@@ -0,0 +1,824 @@
...
+	// cap the delta to 10k vectors
+	if delta > 10_000 {
+		delta = 10_000
+	}
+	var checkpoint uint64
+	if minID > delta {
+		checkpoint = minID - delta
+	} else {
+		checkpoint = 0
+	}
```
[the remark this answers is not in the document]
Comment (asdine) [reply adapters/repos/db/index_queue.go:760]: I added a comment on the function itself, let me know if that's still not clear

File: adapters/repos/db/vector/hnsw/deserializer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (etiennedi) [reply adapters/repos/db/vector/hnsw/deserializer.go]: @trengrj reverted the commit that changed the commit log types, so this is no longer relevant.

**Review by trengrj — APPROVED** [review]

**Comment by msj121** [comment]
With merge does this still need `ASYNC_INDEXING=true`? Also to alleviate memory issues, is there a way to force or allow more of the graph to be stored in harddrive (like qdrant) using the mmap that Weaviate has?

**Pull request merged**